### PR TITLE
add `recursive` option to JSON log format

### LINKED ISSUE #7: RecursiveJSON format?
If the user logs something JSON-able, it might be nice to recurse into that. E.g.

```julia
julia> global_logger(FormatLogger(LoggingFormats.JSON(), stderr))
FormatLogger(LoggingFormats.JSON(), Base.TTY(RawFD(15) open, 0 bytes waiting), true)

julia> @info "Hi" my_json = Dict("hi1" => Dict("hi2" => "bye"))
{"level":"info","msg":"Hi","module":"Main","file":"REPL[86]","line":1,"group":"REPL[86]","id":"Main_b58745e1","kwargs":{"my_json":"Dict(\"hi1\" => Dict(\"hi2\" => \"bye\"))"}}
```
could emit
```julia
julia> @info "Hi" my_json = Dict("hi1" => Dict("hi2" => "bye"))
{"level":"info","msg":"Hi","module":"Main","file":"REPL[90]","line":1,"group":"REPL[90]","id":"Main_b58745e3","kwargs":"my_json":{"hi1":{"hi2":"bye"}}}
```

It's less safe though, since
```julia
julia> struct A end

julia> @info "Hi" my_json = Dict("hi1" => Dict("hi2" => A()))
ERROR: ArgumentError: A doesn't have a defined `StructTypes.StructType`
```
which is probably why you didn't do this, but maybe it could be useful as a separate `RecursiveJSON` format?

(might also be a bit less painful with something like https://github.com/JuliaData/StructTypes.jl/issues/61)

### PR DESCRIPTION
**Pull request by ericphanson**
closes #7

### CODE REVIEW

**Review by racinmat — CHANGES_REQUESTED** [review]
It seems it's not passing for Julia 1.0, which would be nice to make working.

**Comment by fredrikekre** [comment]
Perhaps this could be an option via a kwarg? One of the reason to implement these formats as callable structs was to be able to store options and config in them.

**Comment by ericphanson** [comment]
Sure, that makes sense, I'll give that a try here

File: src/LoggingFormats.jl [hunk 1]
```diff
@@ -56,37 +56,55 @@ lvlstr(lvl::Logging.LogLevel) = lvl >= Logging.Error ? "error" :
...
+JSON(; recursive=false) = JSON(recursive)
+
+function (j::JSON)(io, args)
+    if j.recursive
+        logmsg = JSONLogMessage{Any}(args)
+        try
+            JSON3.write(io, logmsg)
+        catch e
+            fallback_msg = JSONLogMessage{String}(args)
+            fallback_msg.kwargs["JSONRecursionError"] = sprint(Base.showerror, e)
```
Comment (ericphanson) [inline src/LoggingFormats.jl]: or something like
```suggestion
            fallback_msg.kwargs["LoggingFormats_json_recursion_error"] = sprint(Base.showerror, e)
```
?
I don't want it being confused with user-defined keys...

Comment (fredrikekre) [reply src/LoggingFormats.jl]: Maybe `"LoggingFormats.FormatError"`

**Review by racinmat — APPROVED** [review]
LGTM

File: README.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ This package is an aggregation of various useful format functions to use with th
 [LoggingExtras](https://github.com/JuliaLogging/LoggingExtras.jl) package.
 
 Currently, the following functors are available:
-- `JSON`: output log events as JSON
+- `JSON`, `RecursiveJSON`: output log events as JSON
```
Comment (fredrikekre) [inline README.md]:
```suggestion
- `JSON`: output log events as JSON
```

**Review by fredrikekre — APPROVED** [review]

**Pull request merged**